resource/cmake/CommonFunctions.cmake

```diff
@@ -0,0 +1,443 @@
+]]
+macro(m_add_subdirectory_if_exists _SUBDIRECTORY_DIR)
+    if (IS_DIRECTORY "${_SUBDIRECTORY_DIR}")
+        add_subdirectory("${_SUBDIRECTORY_DIR}")
+    endif()
+endmacro()
+
+#[[
+
+m_add_doc_if_exist
+----------------
+
+ * Type: macro
+ * Brief: adds the "doc" directory if it exists in the current source
+          directory. This folder is only added if "building doc" is enabled.
+ * Params: no parameters
+ * How to use it:
+
+    m_add_doc_if_exist()
+]]
+macro(m_add_doc_if_exist)
+    if (RTIGATEWAY_ENABLE_DOCS)
+        m_add_subdirectory_if_exists("${CMAKE_CURRENT_SOURCE_DIR}/doc")
+    endif()
+endmacro()
+
+#[[
+
+m_add_test_if_exist
+-----------------
+
+ * Type: macro
+ * Brief: adds the "test" directory if it exists in the current source
+          directory. This folder is only added if testing is enabled.
+ * Params: no parameters
+ * How to use it:
+
+    m_add_test_if_exist()
+]]
+macro(m_add_test_if_exist)
+    if (RTIGATEWAY_ENABLE_TESTS)
+        m_add_subdirectory_if_exists("${CMAKE_CURRENT_SOURCE_DIR}/test")
+    endif()
+endmacro()
+
+#[[
+
+add_plugin
+----------
+
+ * Type: function
+ * Brief: adds the corresponding plugin directories if they are added. These
+          directories are: the main plugin directory, related resources and
+          related examples for that plugin.
+ * Params:
+ ** ENABLE_PLUGIN: boolean that indicates that this plugin is enabled
+ ** PREFIX: plugin prefix
+ ** STAGING_DIR: staging directory name for the plugin
+ ** PLUGIN_DIR_NAME: name that identifies the folder of that plugin
+ ** PLUGIN_TYPE: indicates whether a plugin is an "ADAPTER", a "PROCESSOR" or a
+                 "TRANSFORMATION"
+ * Restrictions: CONNEXTDDS_DIR should be set
+ * How to use it:
+
+    add_plugin(
+        ENABLE_PLUGIN RTIGATEWAY_ENABLE_MODBUS
+        PREFIX "RTI_MODBUS"
+        STAGING_DIR "modbus"
+        PLUGIN_DIR_NAME "modbus"
+        PLUGIN_TYPE "ADAPTER"
+    )
+]]
+function(add_plugin)
+    set(BOOLEANS ENABLE_PLUGIN)
+    set(SINGLE_VALUE_ARGS PREFIX STAGING_DIR PLUGIN_DIR_NAME PLUGIN_TYPE)
+    set(MULTI_VALUE_ARGS)
+
+    cmake_parse_arguments(_COMMON_FUNCTIONS
+        "${BOOLEANS}"
+        "${SINGLE_VALUE_ARGS}"
+        "${MULTI_VALUE_ARGS}"
+        ${ARGN}
+    )
+
+    set(PLUGIN_TYPE_DIR "")
+    if (${_COMMON_FUNCTIONS_PLUGIN_TYPE} STREQUAL "ADAPTER")
+        set(PLUGIN_TYPE_DIR ${ADAPTERS_DIR})
+    elseif(${_COMMON_FUNCTIONS_PLUGIN_TYPE} STREQUAL "PROCESSOR")
+        set(PLUGIN_TYPE_DIR ${PROCESSORS_DIR})
+    elseif(${_COMMON_FUNCTIONS_PLUGIN_TYPE} STREQUAL "TRANSFORMATION")
+        set(PLUGIN_TYPE_DIR ${TRANSFORMATIONS_DIR})
+    else()
+        mesasge(FATAL_ERROR
+            "The plugin ${_COMMON_FUNCTIONS_PLUGIN_DIR_NAME} is not identified
+            as ADAPTER, PROCESSOR nor TRANSFORMATION"
+        )
+    endif()
+
+    if (${_COMMON_FUNCTIONS_ENABLE_PLUGIN})
+        set(RSPLUGIN_PREFIX ${_COMMON_FUNCTIONS_PREFIX})
```

Comment:
Do we actually need to set (and use) `RSPLUGIN_PREFIX`?

I'll probably have an answer once I finish the review, but I ask the question because I think this variable might be a remnant of the earlier (and more convoluted) version of the build system.

Reply:
We use it in some plugins (mqtt, json transformation and forwarding processor) to add some in-code definitions, such as: `${RSPLUGIN_PREFIX}_ENABLE_LOG`, as part of #8 , probably this will be deleted because we will use internal logger mechanisms.